cpp/src/distance/detail/distance_ops/bitwise_hamming.cuh

```diff
@@ -0,0 +1,60 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include <cuda_runtime.h>
+#include <stdint.h>
+
+namespace cuvs::distance::detail::ops {
+
+/**
+ * @brief the Bitwise Hamming distance matrix calculation
+ *  It computes the following equation:
+ *
+ *    c_ij = sum_k popcount(x_ik XOR y_kj)
+ *
+ * where x and y are binary data packed as uint8_t
+ */
+template <typename DataType, typename AccType, typename IdxType>
+struct bitwise_hamming_distance_op {
+  using DataT = DataType;
+  using AccT  = AccType;
+  using IdxT  = IdxType;
+
+  IdxT k;
+
+  bitwise_hamming_distance_op(IdxT k_) noexcept : k(k_) {}
+
+  static constexpr bool use_norms            = false;
+  static constexpr bool expensive_inner_loop = false;
+
+  template <typename Policy>
+  static constexpr size_t shared_mem_size()
+  {
+    return Policy::SmemSize;
+  }
+
+  __device__ __forceinline__ void core(AccT& acc, DataT& x, DataT& y) const
+  {
+    static_assert(std::is_same_v<DataT, uint8_t>, "BitwiseHamming only supports uint8_t");
+    // Ensure proper masking and casting to avoid undefined behavior
+    uint32_t xor_val    = static_cast<uint32_t>(static_cast<uint8_t>(x ^ y));
+    uint32_t masked_val = xor_val & 0xffu;
+    int popcount        = __popc(masked_val);
```

Comment:
This works, but it is inefficient to use `__popc` one byte at a time. Could you create an issue to note that there is a potential improvement if we use 4 byte storage type for x and y?

Reply:
yeah the reason I did this was to avoid separate compilation paths for larger datatypes such as half and uint32. The idea was to dispatch to the corresponding type (uint32 or half) depending on the dimension (whether it is divisible by 4 or by 2 or not). But that would bloat the binary size because the distance op would have to be defined for these additional types.